lib/async_hooks.js

```diff
@@ -133,6 +133,69 @@ function createHook(fns) {
 }
 
 
+// AsyncLocal API //
+
+const locals = [];
+const localsHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    const execRes = executionAsyncResource();
+    // Using var here instead of let because "for (var ...)" is faster than let.
+    // Refs: https://github.com/nodejs/node/pull/30380#issuecomment-552948364
+    for (var i = 0; i < locals.length; i++) {
+      locals[i][kPropagateSymbol](execRes, resource);
+    }
+  }
+});
+
+const kPropagateSymbol = Symbol('propagate');
+
+class AsyncLocal {
+  constructor() {
+    this.symbol = Symbol('async-local');
+    this.enable();
+  }
+
+  [kPropagateSymbol](execRes, initRes) {
+    initRes[this.symbol] = execRes[this.symbol];
```

Comment:
I think this might miss a way to stop context propagation without disabling. Right now, this would continue adding 1 property per resource for each AsyncLocal even if unused right?

Reply:
`asyncLocal.disable()` is the intended way to stop propagation. It would remove the `asyncLocal` from the `locals` list, preventing the propagator method from being called at all. By writing it this way, it's possible to store falsy values and still have them propagate properly, whereas doing a truthy check of `execRes[this.symbol]` would not.